pkg/registry/service/allocator/allocator.go

```diff
@@ -0,0 +1,270 @@
+	ErrMismatchedNetwork  = errors.New("the provided network does not match the current range")
+	ErrAllocationDisabled = errors.New("IP addresses cannot be allocated at this time")
+)
+
+// Range is a contiguous block of IPs that can be allocated atomically.
+type Range struct {
+	net      *net.IPNet
+	base     *big.Int
+	strategy allocateStrategy
+	max      int
+
+	// lock guards the following members
+	lock      sync.Mutex
+	count     int
+	allocated *big.Int
+}
+
+// allocateStrategy is a search strategy in the allocation map for a valid IP.
+type allocateStrategy func(allocated *big.Int, max, count int) (int, error)
+
+// NewCIDRRange creates a Range over a net.IPNet.
+func NewCIDRRange(cidr *net.IPNet) *Range {
+	ones, bits := cidr.Mask.Size()
+	max := int(1) << uint(bits-ones)
+	r := Range{
+		net:      cidr,
+		strategy: randomScanStrategy,
+		base:     bigForIP(cidr.IP),
+		max:      max,
+
+		allocated: big.NewInt(0),
+		count:     0,
+	}
+	return &r
+}
+
+// Free returns the count of IP addresses left in the range.
+func (r *Range) Free() int {
+	r.lock.Lock()
+	defer r.lock.Unlock()
+	return r.max - r.count
+}
+
+// Allocate attempts to reserve the provided IP. ErrNotInRange or
+// ErrAllocated will be returned if the IP is not valid for this range
+// or has already been reserved.  ErrFull will be returned if there
+// are no addresses left.
+func (r *Range) Allocate(ip net.IP) error {
+	if !r.net.Contains(ip) {
+		return ErrNotInRange
+	}
+
+	r.lock.Lock()
+	defer r.lock.Unlock()
+
+	offset := int(big.NewInt(0).Sub(bigForIP(ip), r.base).Uint64())
+	if r.allocated.Bit(offset) == 1 {
+		return ErrAllocated
+	}
+	r.allocated = r.allocated.SetBit(r.allocated, offset, 1)
+	r.count++
+	return nil
+}
+
+// AllocateNext reserves one of the IPs from the pool. ErrFull may
+// be returned if there are no addresses left.
+func (r *Range) AllocateNext() (net.IP, error) {
+	r.lock.Lock()
+	defer r.lock.Unlock()
+
+	next, err := r.strategy(r.allocated, r.max, r.count)
+	if err != nil {
+		return nil, err
+	}
+	r.count++
+	r.allocated = r.allocated.SetBit(r.allocated, next, 1)
+	return addIPOffset(r.base, int64(next)), nil
+}
+
+// Release releases the IP back to the pool. Releasing an
+// unallocated IP or an IP out of the range is a no-op and
+// returns no error.
+func (r *Range) Release(ip net.IP) error {
+	if !r.net.Contains(ip) {
+		return nil
+	}
+
+	r.lock.Lock()
+	defer r.lock.Unlock()
+
+	offset := int(big.NewInt(0).Sub(bigForIP(ip), r.base).Uint64())
+	if r.allocated.Bit(offset) == 0 {
+		return nil
+	}
+	r.allocated = r.allocated.SetBit(r.allocated, offset, 0)
+	r.count--
+	return nil
+}
+
+// Has returns true if the provided IP is already allocated and a call
+// to Allocate(ip) would fail with ErrAllocated.
+func (r *Range) Has(ip net.IP) bool {
+	if !r.net.Contains(ip) {
+		return false
+	}
+
+	r.lock.Lock()
+	defer r.lock.Unlock()
+
+	offset := int(big.NewInt(0).Sub(bigForIP(ip), r.base).Uint64())
```

Comment:
It'd be nice if you pulled this out into a helper function (like addIPOffset) in the three places it's used for the benefit of people reading the code - it took me longer than I'd like to admit to properly parse that this was subtracting base from ip rather than ip from 0.